tests/pos-macros/quoted-with-precise-types.scala

```diff
@@ -0,0 +1,41 @@
+import scala.quoted.*
+
+def withType[T <: AnyKind, U](tpe: Type[T])(body: [X <: T] => Type[X] ?=> U)(using Quotes): U =
+  type X <: T
+  val tpeX: Type[X] = tpe.asInstanceOf[Type[X]]
+  body[X](using tpeX)
+
+def test1(t1: Type[?], t2: Type[? <: Any])(using Quotes) =
+  withType(t1) { [T <: AnyKind] => _ ?=> // TODO remove _ ?=> // Implementation restriction: polymorphic function literals must have a value parameter
+    Type.of[T]
+    Type.show[T]
+  }
+  withType(t2) { [T] => _ ?=> // TODO remove _ ?=>
+    '{ val a: T = ??? }
+    Type.of[T]
+    Type.show[T]
+  }
+  withType(t2):
+    [T] => _ ?=> '{ val a: T = ??? } // TODO remove _ ?=>
+
+def exprWithPreciseType[T, U](expr: Expr[T])(body: [X <: T] => Type[X] ?=> Expr[X] => U)(using Quotes): U =
+  import quotes.reflect.*
+  type X <: T
+  val exprX = expr.asInstanceOf[Expr[X]]
+  val tpeX = expr.asTerm.tpe.asType.asInstanceOf[Type[X]]
+  body[X](using tpeX)(exprX)
```

Comment:
Why not pass `T` to body and avoid the casts?

Reply:
but body is not inlined so I don't understand how this would change anything to the code we generate

Reply:
It is about the code generated in`body`. For example 
```scala
def foo[T: Type](expr: Expr[T])(using Quotes) =  '{ val y: T = $x }
def bar[T: Type](expr: Expr[T])(using Quotes) = 
  exprWithPreciseType(a) { [X] => _ ?=> x => 
    '{ val y: X = $x }
  }
```
```
val a: Expr[Int] = '{ 1 }
foo(a) // generates '{ val y: Int = 1 }
foo[Any](a) // generates '{ val y: Any = 1 } // boxing
foo(b) // generates '{ val y: Any = 1 } // boxing

val b: Expr[Any] = '{ 2 }
bar(a) // generates '{ val y: Int = 2 }
bar[Any](a) // generates '{ val y: Int = 2 }
bar(b) // generates '{ val y: Int = 2 }
```

Other uses involve getting the precise type to inspect it.

In general it is equivalent but will always be more efficient than
```scala
def baz[T: Type](expr: Expr[T])(using Quotes) = 
  expr match
    case '{ $x: t } =>  '{ val y: t = $x }
```

------
Though the main use case will probably be wen we get a term or type from reflection and want to recover its precise type.

```scala
....
exprWithPreciseType(term.asExpr/*: Expr[Any]*/) { [T] => _ ?=> x =>
   '{ val y: T= $x; ... }
}

withType(tpe.asType/*: Type[? <: AnyKind]*/) { [T] => _ ?=> 
  someFunction[T]//(using Type.of[T])
}
```